pkg/testing/multipass/provisioner.go

```diff
@@ -0,0 +1,229 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package multipass
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/elastic-agent/pkg/core/process"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/runner"
+)
+
+const (
+	Ubuntu = "ubuntu"
+)
+
+type provisioner struct {
+	logger runner.Logger
+}
+
+// NewProvisioner creates the multipass provisioner
+func NewProvisioner() runner.InstanceProvisioner {
+	return &provisioner{}
+}
+
+func (p *provisioner) SetLogger(l runner.Logger) {
+	p.logger = l
+}
+
+// Supported returns true if multipass supports this OS.
+//
+// multipass only supports Ubuntu on the same architecture as the running host.
+func (p *provisioner) Supported(os define.OS) bool {
+	if os.Type != define.Linux {
+		return false
+	}
+	if os.Distro != Ubuntu {
+		return false
+	}
+	if os.Version != "20.04" && os.Version != "22.04" {
+		return false
+	}
+	// multipass only supports the same architecture of the host
+	if os.Arch != runtime.GOARCH {
+		return false
+	}
+	return true
+}
+
+func (p *provisioner) Provision(ctx context.Context, batches []runner.OSBatch) ([]runner.Instance, error) {
+	// this don't provision the instances in parallel on purpose
+	// multipass cannot handle it, it either results in instances sharing the same IP address
+	// or some instances stuck in Starting state
+	for _, batch := range batches {
+		err := func(batch runner.OSBatch) error {
+			launchCtx, launchCancel := context.WithTimeout(ctx, 5*time.Minute)
+			defer launchCancel()
+			err := p.launch(launchCtx, batch)
+			if err != nil {
+				return fmt.Errorf("instance %s failed: %w", batch.ID, err)
+			}
+			return nil
+		}(batch)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var results []runner.Instance
+	instances, err := p.list(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, batch := range batches {
+		mi, ok := instances[batch.ID]
+		if !ok {
+			return nil, fmt.Errorf("failed to find %s in multipass list output", batch.ID)
+		}
+		if mi.State != "Running" {
+			return nil, fmt.Errorf("instance %s is not marked as running", batch.ID)
+		}
+		results = append(results, runner.Instance{
+			ID:         batch.ID,
+			Name:       batch.ID,
+			IP:         mi.IPv4[0],
+			Username:   "ubuntu",
+			RemotePath: "/home/ubuntu/agent",
+			Internal:   nil,
+		})
+	}
+	return results, nil
+}
+
+// Clean cleans up all provisioned resources.
+func (p *provisioner) Clean(ctx context.Context, instances []runner.Instance) error {
+	// doesn't execute in parallel for the same reasons in Provision
+	// multipass just cannot handle it
+	for _, instance := range instances {
+		func(instance runner.Instance) {
+			deleteCtx, deleteCancel := context.WithTimeout(ctx, 5*time.Minute)
+			defer deleteCancel()
+			err := p.delete(deleteCtx, instance)
```

Comment:
do you want to ```prune``` as well here?

Reply:
Do you mean `purge`? If so, I do that in the delete call with `multipass delete -p [instanceName]`